supportedBackends/google/pipelines/v2alpha1/src/main/scala/cromwell/backend/google/pipelines/v2alpha1/PipelinesApiAsyncBackendJobExecutionActor.scala

```diff
@@ -39,6 +48,105 @@ class PipelinesApiAsyncBackendJobExecutionActor(standardParams: StandardAsyncExe
       })
   }
 
+  private lazy val transferScriptTemplate =
+    Source.fromInputStream(Thread.currentThread.getContextClassLoader.getResourceAsStream("transfer.sh")).mkString
+
+  private def localizationTransferBundle[T <: PipelinesApiInput](localizationConfiguration: LocalizationConfiguration, files: Boolean)(bucket: String, inputs: List[T]): String = {
+    // `.head` is safe as there is always at least one input or this method wouldn't be invoked for the specified bucket.
+    val project = inputs.head.cloudPath.asInstanceOf[GcsPath].projectId
+    val maxAttempts = localizationConfiguration.localizationAttempts
+    val cloudAndContainerPaths = inputs.flatMap { i => List(i.cloudPath, i.containerPath) } mkString("\"", "\"\n|  \"", "\"")
+
+    val filesOrDirectories = if (files) "files" else "directories"
+
+    // Use a digest as bucket names can contain characters that are not legal in bash identifiers.
+    val arrayIdentifier = s"localize_${filesOrDirectories}_" + DigestUtils.md5Hex(bucket).take(7)
+    s"""
+       |# $bucket
+       |$arrayIdentifier=(
+       |  "localize" # direction
+       |  "$filesOrDirectories"    # files or directories
+       |  "$project"       # project
+       |  "$maxAttempts"   # max attempts
+       |  $cloudAndContainerPaths
+       |)
+       |
+       |transfer "$${$arrayIdentifier[@]}"
+      """.stripMargin
+  }
+
+  private def delocalizationTransferBundle[T <: PipelinesApiOutput](localizationConfiguration: LocalizationConfiguration, files: Boolean)(bucket: String, outputs: List[T]): String = {
+    // `.head` is safe as there is always at least one output or this method wouldn't be invoked for the specified bucket.
+    val project = outputs.head.cloudPath.asInstanceOf[GcsPath].projectId
+    val maxAttempts = localizationConfiguration.localizationAttempts
+    val cloudAndContainerPaths = outputs.flatMap { o => List(o.cloudPath, o.containerPath, o.contentType.getOrElse("")) } mkString("\"", "\"\n|  \"", "\"")
+
+    val filesOrDirectories = if (files) "files" else "directories"
+
+    // Use a digest as bucket names can contain characters that are not legal in bash identifiers.
+    val arrayIdentifier = s"delocalize_${filesOrDirectories}_" + DigestUtils.md5Hex(bucket).take(7)
+    s"""
+       |# $bucket
+       |$arrayIdentifier=(
+       |  "delocalize" # direction
+       |  "$filesOrDirectories"      # files or directories
+       |  "$project"       # project
+       |  "$maxAttempts"   # max attempts
+       |  $cloudAndContainerPaths
+       |)
+       |
+       |transfer "$${$arrayIdentifier[@]}"
+      """.stripMargin
+  }
+
+  private def generateLocalizationScript(inputs: List[PipelinesApiInput], mounts: List[Mount])(implicit localizationConfiguration: LocalizationConfiguration): String = {
```

Comment:
One caution: if you have a very large number of inputs you might exceed the maximum allowed command line length and cause the script invocation to fail.  It might be necessary to break these command lines into chunks, depending on how many files you expect to localize per bucket.

Reply:
Oh we have an integration test with more enough inputs to run into that problem, that's why this script is localized to GCS and then delocalized to the node and executed. 🙂

Reply:
I may be missing something since I don't know scala at all, but the command line limit would still apply if you are ever calling transfer.sh with too many arguments: so the script itself gets localized but transfer.sh is still invoked with an argument list, correct?  eg, something like

  `transfer.sh <list of files> <list of directories>`

If that exceeds the system's ARG_MAX it will break (without using, eg xargs).  I think the value for Container Optimized OS is something on the order of 2MB so you may not have hit this limit yet and it may not be a concern at the moment.

Reply:
Oh the script will be invoked with no arguments at all. `transfer.sh` is the preamble for what becomes `gcs_localization` or `gcs_delocalization` on the node and will be executed something like `/bin/bash /cromwell_root/gcs_localization`. Bash arrays are appended to the `transfer.sh` preamble containing all information required to transfer files and directories per bucket. Something like:

```
<transfer.sh preamble>
...
# 12345 would be a bucket's digest
localize_12345=(
  ...
)

# invoke the `transfer` bash function in transfer.sh passing the array for the first bucket
transfer "${localize_12345[@]}"

# localize for all the other buckets
localize_67890=(
...
)
...
```

Reply:
Got it, thanks!